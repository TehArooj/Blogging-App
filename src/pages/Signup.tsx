import { FormEvent, useState } from "react";
import { Link } from "react-router-dom";
import { createUserWithEmailAndPassword, updateProfile } from "firebase/auth";
import CustomInput from "../components/CustomInput.component";
import { auth, db } from "../utils/firebase/firebase.utils";
import { useNavigate } from "react-router-dom";
import { doc, getDoc, setDoc } from "firebase/firestore";

const Signup = () => {
  const navigate = useNavigate();

  const [values, setValues] = useState({
    fullName: "",
    email: "",
    password: "",
    confirmPassword: "",
  });

  const [errorMsg, setErrorMsg] = useState("");
  const [submitButtonDisabled, setSubmitButtonDisabled] = useState(false);

  const handleChangeName = (e: FormEvent) => {
    e.preventDefault();
    setValues((prev) => ({
      ...prev,
      fullName: (e.target as HTMLInputElement).value,
    }));
  };

  const handleChangeEmail = (e: FormEvent) => {
    e.preventDefault();
    setValues((prev) => ({
      ...prev,
      email: (e.target as HTMLInputElement).value,
    }));
  };

  const handleChangePassword = (e: FormEvent) => {
    e.preventDefault();
    setValues((prev) => ({
      ...prev,
      password: (e.target as HTMLInputElement).value,
    }));
  };

  const handleChangeConfirmPassword = (e: FormEvent) => {
    e.preventDefault();
    setValues((prev) => ({
      ...prev,
      confirmPassword: (e.target as HTMLInputElement).value,
    }));
  };

  const isValidEmail = (email: string) => {
    return /\S+@\S+\.\S+/.test(email);
  };

  const validations = () => {
    if (
      !values.fullName ||
      !values.email ||
      !values.password ||
      !values.confirmPassword
    ) {
      setErrorMsg("Please fill all the fields.");
      return;
    }
    setErrorMsg("");

    if (!isValidEmail(values.email)) {
      setErrorMsg("Invalid Email.");
      return;
    } else {
      setErrorMsg("");
    }

    values.password.length < 6
      ? setErrorMsg("Password must be atleast six characters.")
      : values.password !== values.confirmPassword
      ? setErrorMsg("Passwords didn't match please try again.")
      : setErrorMsg("");
  };

  const handleSubmission = (e: FormEvent) => {
    e.preventDefault();

    // Validations
    validations();

    setSubmitButtonDisabled(true);

    //Sign up user
    createUserWithEmailAndPassword(auth, values.email, values.confirmPassword)
      .then(async (res) => {
        setSubmitButtonDisabled(false);
        const user = res.user;
        await updateProfile(user, { displayName: values.fullName });

        if (!user) return;

        const userDocRef = doc(db, "users", user.uid);

        const userSnapshot = await getDoc(userDocRef);

        if (!userSnapshot.exists()) {
          const { displayName, email } = user;
          const createdAt = new Date();

          try {
            await setDoc(userDocRef, {
              displayName,
              email,
              createdAt,
            });
          } catch (error) {
            console.log("error creating the user", error);
          }
          console.log(user);
          navigate("/home");
        }
      })
      .catch((err) => {
        setSubmitButtonDisabled(false);
      });
  };

  return (
    <div className=" grid grid-cols-3 gap-4 justify-evenly not-italic w-full h-full md:grid-cols-4 tb:grid-cols-1 m:grid-cols-1">
      <div className="flex items-center justify-center w-full h-full md:col-span-2 ">
        <div className="flex items-center justify-center font-lexend absolute w-96 h-full left-0 top-0  bg-pawel-pattern  bg-cover object-fill bg-no-repeat tb:hidden m:hidden">
          <div className="inherit text-6xl text-white text-center rotate-270 font-bold ">
            Sign Up
          </div>
        </div>
      </div>
      <div className="col-span-2 flex items-center justify-center md:col-span-2 md:justify-center tb:col-span-auto tb:justify-center  m:justify-center">
        <div className=" mt-36 left-1/2 lg:ml-24 lg:mr-20 md:ml-8 md:mr-8 tb:items-center tb:justify-center tb:ml-48 tb:mr-48 tb:mt-24 m:ml-36 m:mr-36 m:mt-16 m:items-center m:justify-center ">
          <h1 className=" font-dm font-bold text-5xl  text-left  text-darkGrey tb:text-center m:text-center">
            Welcome
          </h1>
          <div className="font-lexend flex flex-col ">
            <form>
              <div className="text-secondary ">
                <p className="text-2xl font-light mb-5 tb:text-center m:text-center m:text-xl ">
                  Let's sign you up quickly
                </p>
                <CustomInput
                  type="text"
                  placeholder="Full Name"
                  handleChange={handleChangeName}
                />
                <CustomInput
                  type="text"
                  placeholder="Email Address"
                  handleChange={handleChangeEmail}
                />
                <CustomInput
                  type="password"
                  placeholder="Password"
                  handleChange={handleChangePassword}
                />
                <CustomInput
                  type="password"
                  placeholder="Confirm Password"
                  handleChange={handleChangeConfirmPassword}
                />
              </div>
              <div className="mb-5">
                <b className="text-base text-errorMsg m:text-sm ">{errorMsg}</b>
              </div>
              <button
                className="text-white font-semibold bg-secondary border-solid border-2  border-secondary h-14  w-44 mb-9 hover:outline-none hover:bg-darkGrey hover:border-none  disabled:bg-gray-500 md:w-80 tb:w-96 tb:h-[52px] m:w-72 m:h-[52px] m:text-sm"
                type="submit"
                onClick={handleSubmission}
                disabled={submitButtonDisabled}
              >
                SUBMIT
              </button>
            </form>
            <div className="text-xl  tb:text-center m:text-base m:text-center">
              <p className="text-darkGrey">
                Already have an account?
                <Link to="/login" className="text-primary  ml-1">
                  Log-in
                </Link>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Signup;
