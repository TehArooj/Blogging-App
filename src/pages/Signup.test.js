import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import Signup from "./Signup";
import CustomInput from "../components/CustomInput.component";
import LoaderSpinner from "../components/LoaderSpinner.component";

it("<Signup /> renders correctly", async () => {
  const tree = renderer
    .create(
      <Router>
        <Signup />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("<Signup /> renders correctly when there is a custom input", () => {
  const tree = renderer.create(<CustomInput />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("<Signup /> renders correctly when there is custom input with props", () => {
  const tree = renderer
    .create(
      <CustomInput placeholder="hello" type="text" handleChange={() => {}} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("<Signup /> renders correctly when there is a Loader Spinner", () => {
  const tree = renderer.create(<LoaderSpinner />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("<Signup /> check matchers and pass", async () => {
  const user = {
    fullname:"tester",
    email: "test@gmail.com",
    password: "123456",
    confirmPassword:"123456",
  };

  expect(user).toMatchSnapshot({
    fullname: expect.any(String),
    email: expect.any(String),
    password: "123456",
    confirmPassword: "123456",
  });
});