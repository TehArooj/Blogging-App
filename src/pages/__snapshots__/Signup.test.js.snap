// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> check matchers and pass 1`] = `
Object {
  "confirmPassword": "123456",
  "email": Any<String>,
  "fullname": Any<String>,
  "password": "123456",
}
`;

exports[`<Signup /> renders correctly 1`] = `
<div
  className=" flex not-italic w-full h-full"
>
  <div
    className="flex items-start justify-center w-[426px] lg:w-[400px]  md:w-[380px] h-full tb:hidden m:hidden"
  >
    <div
      className="flex items-center justify-center font-lexend w-[426px] lg:w-[400px] md:w-[380px]  h-screen left-0 top-0  bg-pawel-pattern  bg-cover object-fill bg-no-repeat tb:hidden m:hidden"
    >
      <div
        className="inherit text-6xl text-white text-center rotate-270  font-bold "
      >
        Sign Up
      </div>
    </div>
  </div>
  <div
    className="flex flex-col w-full tb:w-full m:w-full h-full"
  >
    <div
      className=" mt-36 2xl:pl-[100px] xl:pl-24 xl:pr-24 lg:mr-0 md:ml-0 md:mr-0 lg:ml-0 lg:pr-32 lg:pl-32 md:pr-20 md:pl-20 tb:pl-10 tb:pr-10 tb:items-center tb:justify-center tb:ml-0 tb:mr-0  tb:mt-24 m:ml-0 m:mr-0 m:pl-8 m:pr-8 m:mt-24 m:items-center m:justify-center "
    >
      <h1
        className=" font-dm font-bold text-5xl  text-left  text-darkGrey tb:text-center m:text-center"
      >
        Welcome
      </h1>
      <div
        className="font-lexend flex flex-col"
      >
        <form>
          <div
            className="text-secondary"
          >
            <p
              className="text-2xl font-light mb-5 tb:text-center m:text-center m:text-xl"
            >
              Let's sign you up quickly
            </p>
            <input
              className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
              onChange={[Function]}
              placeholder="Full Name"
              required={true}
              type="text"
            />
            <input
              className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
              onChange={[Function]}
              placeholder="Email Address"
              required={true}
              type="text"
            />
            <input
              className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
            />
            <input
              className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
              onChange={[Function]}
              placeholder="Confirm Password"
              required={true}
              type="password"
            />
          </div>
          <div
            className="mb-5"
          >
            <b
              className="text-base text-errorMsg m:text-sm "
            />
          </div>
          <button
            className="text-white font-semibold bg-secondary border-solid border-2  border-secondary h-14  w-44 mb-9 hover:outline-none hover:bg-darkGrey hover:border-none  disabled:bg-gray-500  md:w-full tb:w-full tb:h-[52px] m:w-full m:h-[52px] m:text-sm"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            SUBMIT
          </button>
        </form>
        <div
          className="text-xl  tb:text-center m:text-base m:text-center"
        >
          <p
            className="text-darkGrey"
          >
            Already have an account?
            <a
              className="text-primary  ml-1"
              href="/login"
              onClick={[Function]}
            >
              Log-in
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Signup /> renders correctly when there is a Loader Spinner 1`] = `
<div
  className="absolute top-[45%] left-[45%] m:top-[40%] m:left-[40%]"
>
  <div
    aria-busy={true}
    aria-label="ball-triangle-loading"
    data-testid="ball-triangle-loading"
    role="status"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <svg
      data-testid="ball-triangle-svg"
      height={100}
      stroke="#4fa94d"
      viewBox="0 0 57 57"
      width={100}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fillRule="evenodd"
      >
        <g
          strokeWidth="2"
          transform="translate(1 1)"
        >
          <circle
            cx="5"
            cy="50"
            r={5}
          >
            <animate
              attributeName="cy"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              repeatCount="indefinite"
              values="50;5;50;50"
            />
            <animate
              attributeName="cx"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              repeatCount="indefinite"
              values="5;27;49;5"
            />
          </circle>
          <circle
            cx="27"
            cy="5"
            r={5}
          >
            <animate
              attributeName="cy"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              from="5"
              repeatCount="indefinite"
              to="5"
              values="5;50;50;5"
            />
            <animate
              attributeName="cx"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              from="27"
              repeatCount="indefinite"
              to="27"
              values="27;49;5;27"
            />
          </circle>
          <circle
            cx="49"
            cy="50"
            r={5}
          >
            <animate
              attributeName="cy"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              repeatCount="indefinite"
              values="50;50;5;50"
            />
            <animate
              attributeName="cx"
              begin="0s"
              calcMode="linear"
              dur="2.2s"
              from="49"
              repeatCount="indefinite"
              to="49"
              values="49;5;27;49"
            />
          </circle>
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`<Signup /> renders correctly when there is a custom input 1`] = `
<input
  className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
  required={true}
/>
`;

exports[`<Signup /> renders correctly when there is custom input with props 1`] = `
<input
  className="text-darkGrey block border-solid border-2 border-secondary pl-8 mb-5 w-2/3 xl:w-full h-[60px] focus:outline-none focus:border-primary lg:w-full md:w-full tb:w-full tb:pl-4 tb:text-sm m:w-full m:text-sm m:pl-4 "
  onChange={[Function]}
  placeholder="hello"
  required={true}
  type="text"
/>
`;
